File: hw/jacome/U3/HW21TemplateMethod/JS/model/CaffeineBeverage.js
class CaffeineBeverage {
    prepareRecipe() {
        this.boilWater();
        this.brew();
        this.pourInCup();
        // Eliminamos la condición de wantsCondiments de aquí
    }
    
    boilWater() {
        console.log("Boiling water");
    }
    
    brew() {
        throw new Error("Abstract method brew() must be implemented");
    }
    
    pourInCup() {
        console.log("Pouring into cup");
    }
    
    addCondiments() {
        throw new Error("Abstract method addCondiments() must be implemented");
    }
}

module.exports = CaffeineBeverage;

File: hw/jacome/U3/HW21TemplateMethod/HW21TemplateMethod/src/main/java/ec/edu/espe/hw21templatemethod/model/CaffeineBeverage.java
/*
 * Click nbfs://nbhost/SystemFileSystem/Templates/Licenses/license-default.txt to change this license
 * Click nbfs://nbhost/SystemFileSystem/Templates/Classes/Class.java to edit this template
 */

package ec.edu.espe.hw21templatemethod.model;

/**
 *
 * @author Micaela Jácome
 */
public abstract class CaffeineBeverage {
    public void prepareRecipe(){
        boilWater();
        brew();
        pourInCup();
        if(wantsCondiments()){
            addCondiments();
        }
    }
    
    void boilWater(){
        System.out.println("Boiling water");
    }
    
    abstract void brew();
    
    void pourInCup(){
        System.out.println("Pouring into cup");
    }
    
    abstract void addCondiments();
    
    boolean wantsCondiments(){
        return true;
    }
    
}

